psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1498 @@
+ *
+ * Copyright (c) 2007-2021 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "mp2.h"
+#include "sparse.h"
+
+#include "psi4/psi4-dec.h"
+#include "psi4/physconst.h"
+#include "psi4/psifiles.h"
+
+#include "psi4/lib3index/3index.h"
+#include "psi4/libdiis/diismanager.h"
+#include "psi4/libfock/apps.h"
+#include "psi4/libfock/jk.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/libfock/points.h"
+#include "psi4/libfock/v.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/extern.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/libmints/local.h"
+#include "psi4/libmints/matrix.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/oeprop.h"
+#include "psi4/libmints/twobody.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libpsi4util/PsiOutStream.h"
+#include "psi4/libpsi4util/process.h"
+#include "psi4/libpsio/psio.h"
+#include "psi4/libpsio/psio.hpp"
+#include "psi4/libqt/qt.h"
+#include "psi4/libciomr/libciomr.h"
+
+#include <algorithm>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+
+namespace psi {
+namespace dlpno {
+
+DLPNOMP2::DLPNOMP2(SharedWavefunction ref_wfn, Options& options) : Wavefunction(options) {
+    shallow_copy(ref_wfn);
+    reference_wavefunction_ = ref_wfn;
+
+    common_init();
+}
+DLPNOMP2::~DLPNOMP2() {}
+void DLPNOMP2::common_init() {
+    print_ = options_.get_int("PRINT");
+    debug_ = options_.get_int("DEBUG");
+
+    name_ = "DLPNO-MP2";
+    module_ = "dlpno";
+
+    variables_["SCF TOTAL ENERGY"] = reference_wavefunction_->energy();
+
+    ribasis_ = get_basisset("DF_BASIS_MP2");
+}
+
+/* Utility function for easily performing making C_DGESV calls
+ *
+ * C_DGESV solves AX=B for X, given symmetric NxN matrix A and NxM matrix B
+ * B is expected in fortran layout, which complicates the call when (M > 1)
+ * The workaround used here is switching layouts of B before and after the call
+ */
+void C_DGESV_wrapper(SharedMatrix A, SharedMatrix B) {
+
+    int N = B->rowspi(0);
+    int M = B->colspi(0);
+    if(N == 0 || M == 0) return;
+
+    // create a copy of B in fortran ordering
+    double* B_fortran = new double[N * M];
```

Comment:
You're leaking memory, since you're not deleting the array. Use `unique_ptr` or switch to `std::vector`.